### PR TITLE
feat: Migrate LogSchema::message_key to new lookup code

### PR DESCRIPTION
**Pull request by pront**
## Motivation
This part of https://github.com/vectordotdev/vector/issues/13033.

## Summary
* `LogSchema::message_key` is now an `OptionalValuePath`.
* To avoid hacky `String` to `&'static str` conversions, I changed the `Requirement::meaning` key type to `String`.

### CODE REVIEW

File: lib/vector-core/src/event/vrl_target.rs
Comment (spencergilbert) [inline lib/vector-core/src/event/vrl_target.rs:55]: **🥜 nitpick:** ‏ Looking at the other `match` arms below passing `value` into the function first would match up better.

File: lib/vector-core/src/config/log_schema.rs
Comment (spencergilbert) [inline lib/vector-core/src/config/log_schema.rs:116]: **❓ question:** ‏ I don't understand why this was renamed?

**Review by spencergilbert — DISMISSED** [review]

**Review by spencergilbert — APPROVED** [review]

File: lib/vector-core/src/config/log_schema.rs
Comment (bruceg) [inline lib/vector-core/src/config/log_schema.rs:87]: Should these be in a `const` elsewhere?

File: lib/vector-core/src/event/log_event.rs
Comment (bruceg) [inline lib/vector-core/src/event/log_event.rs:155]: This pattern to looks to be repeated all over. Would an insert helper be useful, maybe something like this?
```rust
    fn maybe_insert(&mut self, prefix: PathPrefix, path: OptionalValuePath, value: impl Into<Value>)
```
(the `impl Into` is to avoid calling `.into` unless the path is `Some`)

File: lib/vector-core/src/schema/requirement.rs
Comment (bruceg) [inline lib/vector-core/src/schema/requirement.rs:17]: Would it be useful to use a `Cow<'static, str>` to support both here? Is this a mapping that is manipulated frequently, or just created once and looked up frequently?

File: src/sources/docker_logs/mod.rs
Comment (bruceg) [inline src/sources/docker_logs/mod.rs:1169]: Am I right to presume this `unwrap` is safe because of the legacy namespace? It might be helpful to document this with a comment or by using `expect` instead (here and below).

File: src/transforms/dedupe.rs
Comment (bruceg) [inline src/transforms/dedupe.rs:106]: Prefer avoiding the overhead of the macro for empty arrays:
```suggestion
    let mut fields = Vec::new();
```

File: lib/codecs/src/encoding/format/raw_message.rs
Comment (bruceg) [inline lib/codecs/src/encoding/format/raw_message.rs:61]: This seems line an expensive operation (`message_key.to_string()`) to be doing on each event encoding. Could it be amortized elsewhere?

File: lib/codecs/src/encoding/format/text.rs
Comment (bruceg) [inline lib/codecs/src/encoding/format/text.rs:79]: Ditto expensive operation here

File: lib/vector-core/src/event/log_event.rs
Comment (pront) [reply lib/vector-core/src/event/log_event.rs]: Added `maybe_insert` function. Saved a few lines.

File: src/sources/docker_logs/mod.rs
Comment (pront) [reply src/sources/docker_logs/mod.rs]: Replaced with `expect(..)`. Yes, message key should not be `None` at this point and (same goes for other instances in this file).

**Review by pront — COMMENTED** [review]
Thanks for the review @bruceg!

I will followup on the `Cow<>` vs `String` question and the performance concerns soon.

File: lib/codecs/src/encoding/format/raw_message.rs
Comment (pront) [reply lib/codecs/src/encoding/format/raw_message.rs]: Excellent catch. We can improve the existing code and avoid expensive copies by using [get_message()](https://github.com/vectordotdev/vector/blob/master/lib/vector-core/src/event/log_event.rs#L478). This is efficient both Vector and Legacy namespaces.

File: lib/codecs/src/encoding/format/text.rs
Comment (pront) [reply lib/codecs/src/encoding/format/text.rs]: Now using `get_message` as above.

File: lib/vector-core/src/schema/requirement.rs
Comment (pront) [reply lib/vector-core/src/schema/requirement.rs:17]: Based on a quick audit, I don't see this being in any critical path. Specifically, it is used to instantiate `schema::Requirement` structs. Then, it is looked up only in one [place](https://github.com/vectordotdev/vector/blob/master/src/topology/schema.rs#L326). I _assume_ this won't have any significant performance impact. 

Regardless, happy to refactor this to `Cow`. Let me know what you think.

File: lib/vector-core/src/config/log_schema.rs
Comment (pront) [reply lib/vector-core/src/config/log_schema.rs:87]: Are you referring to the hardcoded string or `OptionalValuePath`? Happy to refactor this (and the other default_* functions).

File: lib/vector-core/src/schema/requirement.rs
Comment (bruceg) [reply lib/vector-core/src/schema/requirement.rs:17]: No, if it's not critical path then do whatever is simplest (which is _not_ `Cow`).

File: lib/vector-core/src/config/log_schema.rs
Comment (bruceg) [reply lib/vector-core/src/config/log_schema.rs:87]: The hard coded string constant. We just added a style doc note about using `const` values for those.

File: lib/vector-core/src/event/log_event.rs
Comment (bruceg) [inline lib/vector-core/src/event/log_event.rs:158]: For future reference: This pattern could now use the same `maybe_insert` helper, except that we would have to pass in `Utc::now()` unconditionally, which means grabbing system time even when there is no timestamp key. This is out of scope for the current PR but something to think about. You'd need to change `maybe_insert` to take a closure instead of an `Into`, so that adds complication.

File: lib/vector-core/src/event/log_event.rs
Comment (bruceg) [inline lib/vector-core/src/event/log_event.rs:154]: suggestion: Now that you've went and refactored all these, it might be worth adding another helper to specifically encode this one pattern into something like `log.maybe_insert_message(msg)`

**Review by bruceg — DISMISSED** [review]
I left a couple of non-blocking comments, but this looks good as is.

File: lib/vector-core/src/event/log_event.rs
Comment (pront) [reply lib/vector-core/src/event/log_event.rs:158]: Yes, good point. There might be more cases like this with other LogSchema fields. My plan was/is to follow-up with a smaller PR because this one is getting too complex.

**Comment by pront** [comment]
Hi all, 

This PR got a couple of thorough reviews but it still requires several approvals (see [CODEOWNERS](https://github.com/vectordotdev/vector/blob/master/.github/CODEOWNERS)). To get this merged (before many conflicts arise) I would greatly appreciate a review from the following people: 

- @jszwedko: from the standpoint of the processing team (since Nathan is on leave)
- @StephenWakely: for files you own
- @neuronull: for files you own
- @dsmith3197: for files you own 

Friendly reminder that each new commit will invalidate all previous stamps and unfortunately I will have to ping you again for a new review. If there are non-blocking concerns/ideas after the reviews, I am happy to follow up with more PRs. 

Thanks!

**Review by neuronull — APPROVED** [review]
LGTM- I checked mostly just my codeowner files and the rollout to the components seems procedural and straight forward.

**Review by bruceg — DISMISSED** [review]

**Review by dsmith3197 — APPROVED** [review]

**Review by StephenWakely — DISMISSED** [review]
This look good. 

But, there are a number of places where it will panic if the config attempts to turn off the message field. It isn't a regression, it would have panicked before. I can't think of a single useful scenario where someone would want to turn off the message field, so I do wonder if we should actually disallow this for `message_key` and then set 

```rust
 message_key: OwnedValuePath,
```

**Comment by pront** [comment]
> This look good.
> 
> But, there are a number of places where it will panic if the config attempts to turn off the message field. It isn't a regression, it would have panicked before. I can't think of a single useful scenario where someone would want to turn off the message field, so I do wonder if we should actually disallow this for `message_key` and then set
> 
> ```rust
>  message_key: OwnedValuePath,
> ```

Well spotted. I didn't want to introduce any breaking changes here. We can discuss offline and potentially change this.

**Review by jszwedko — APPROVED** [review]

**Review by dsmith3197 — APPROVED** [review]

**Pull request merged**
